test/e2e/dra/dra.go

```diff
@@ -281,6 +285,81 @@ var _ = framework.SIGDescribe("node")("DRA", feature.DynamicResourceAllocation,
+			// is not necessarily done yet with tearning down the
+			// pod at the time when the pod and its claim are
+			// already removed. The user can replace the claim and
+			// pod with new instances under the same name.  The
+			// kubelet then needs to detect that the new claim is
+			// not the same as the one that kubelet currently works
+			// on and that the new pod cannot start until the old
+			// one is torn down.
+			//
+			// This test delays termination of the first pod to ensure
+			// that the race goes bad (old pod pending when new one
+			// arrives) and always schedules to the same node.
+			claim := b.externalClaim()
+			pod := b.podExternal()
+			node := nodes.NodeNames[0]
+			pod.Spec.NodeSelector = map[string]string{"kubernetes.io/hostname": node}
+			oldClaim := b.create(ctx, claim, pod)[0].(*resourceapi.ResourceClaim)
+			b.testPod(ctx, f, pod)
+
+			ginkgo.By("Force-delete claim and pod")
+			forceDelete := metav1.DeleteOptions{GracePeriodSeconds: ptr.To(int64(0))}
+			framework.ExpectNoError(b.f.ClientSet.CoreV1().Pods(b.f.Namespace.Name).Delete(ctx, pod.Name, forceDelete))
+
+			// Fail NodeUnprepareResources to simulate long grace period
+			unprepareResources := MethodInstance{node, NodeUnprepareResourcesMethod}
+			driver.Fail(unprepareResources, true)
+
+			// The pod should get deleted immediately.
+			_, err := b.f.ClientSet.CoreV1().Pods(b.f.Namespace.Name).Get(ctx, pod.Name, metav1.GetOptions{})
+			if !apierrors.IsNotFound(err) {
+				framework.Failf("Expected 'not found' error, got: %v", err)
+			}
+
+			// The claim may take a bit longer because of the allocation and finalizer.
+			framework.ExpectNoError(b.f.ClientSet.ResourceV1beta1().ResourceClaims(b.f.Namespace.Name).Delete(ctx, claim.Name, forceDelete))
+			gomega.Eventually(ctx, func(ctx context.Context) (*resourceapi.ResourceClaim, error) {
+				claim, err := b.f.ClientSet.ResourceV1beta2().ResourceClaims(b.f.Namespace.Name).Get(ctx, claim.Name, metav1.GetOptions{})
+				if apierrors.IsNotFound(err) {
+					return nil, nil
+				}
+				return claim, err
+			}).Should(gomega.BeNil())
+			gomega.Expect(b.driver.Nodes[node].GetPreparedResources()).Should(gomega.Equal([]testdriverapp.ClaimID{{Name: oldClaim.Name, UID: oldClaim.UID}}), "Old claim should still be prepared.")
+
+			ginkgo.By("Re-creating the same claim and pod")
+			newClaim := b.create(ctx, claim, pod)[0].(*resourceapi.ResourceClaim)
+
+			// Keep blocking NodeUnprepareResources for the old pod
+			// until the new pod calls NodePrepareResources and fails.
+			// This ensures that the race is triggered.
+			expectedEvent := fields.Set{
+				"involvedObject.kind":      "Pod",
+				"involvedObject.name":      pod.Name,
+				"involvedObject.namespace": pod.Namespace,
+				"reason":                   events.FailedPrepareDynamicResources,
+			}.AsSelector().String()
+			gomega.Eventually(ctx, func(ctx context.Context) error {
+				return e2eevents.WaitTimeoutForEvent(
```

Comment:
Why are you wrapping a function which polls itself in `gomega.Eventually`? Wouldn't it be simpler to call `e2eevents.WaitTimeoutForEvent` once with the overall timeout that you want?

Speaking of that timeout, what we want here is something that is long enough that the kubelet should have emitted the event multiple times (in case that we missed an event). `3 * f.Timeouts.PodDelete`? A comment about the choice of the timeout for posterity would be good.